Mathlib/Data/Fin/Basic.lean

```diff
@@ -779,6 +780,7 @@ theorem zero_ne_one' [NeZero n] : (0 : Fin (n + 1)) ≠ 1 := Fin.ne_of_lt one_po
 
 #align fin.succ_zero_eq_one' Fin.succ_zero_eq_one
 
+unseal Nat.modCore in
```

Comment:
I'm not particularly happy with this. How would I discover the need to add this unseal, after seeing the `rfl` fail on `@Eq (Fin (n✝ + 1 + 1).succ) (succ 1) 2`?

Reply:
I'm not happy either. Any ideas?

We could just make `Nat.modCore` semireducible again, of course.

Or try harder to make `Nat.mod` reduce when the first argument is any fully known value and it's reducibly smaller than the second.

Reply:
Or use a structurally recursive definition of `Nat.mod` (with fuel)

Reply:
Ah, I think I found the snag why I couldn’t make that work initially. I tried

```
def mod' (n m : Nat) := match n with
  | 0 => 0
  | _ => if n < m then n else Nat.modCore n m
```
but it wouldn’t reduce where I wanted it to.

I now see why. Note that
```
example : decide (m+3 ≤ 2) = false := rfl -- does not work
example : decide (2 < m+3) = false := rfl -- works
```

So 
```
def mod' (n m : Nat) := match n with
  | 0 => 0
  | _ => if m ≤ n then Nat.modCore n m else n
```
might do what we want (namely reduce when `n` is `0`, or when `n` is known and `m` is known enough.) Will try.

Reply:
Thanks for nudging me to try harder, I think I found the right magic incarnation to make this work:
https://github.com/leanprover/lean4/pull/4145

Reply:
Okay, we will wait on this PR until nightly-2024-05-14 lands containing https://github.com/leanprover/lean4/pull/4145!